Refactor ProcessManager to Use ProcessBuilder

mz1999 opened the pull request:
This PR aims to refactor the `ProcessManager`, transitioning from the usage of `Runtime.exec()` to `ProcessBuilder`. The main objective is to provide better control and configurability over the process execution. 

I have also added some tests for `ProcessManager`.

Based on the suggestions in [issue 23565](https://github.com/eclipse-ee4j/glassfish/issues/23565), I had planned to use the refactored `ProcessManager` instead of `ProcessExecutor`, but I found that I could not simply replace it. The `ProcessExecutor` redirects the standard output to a file, which is different from the mechanism of `ProcessManager` and requires more careful consideration and testing, so I still finished the refactoring of `ProcessManager` first.

mz1999 committed:
Refactor ProcessManager to Use ProcessBuilder

avpinchuk commented:
@mz1999, I think you should suggest the second commit as a separate PR

mz1999 commented:
> @mz1999, I think you should suggest the second commit as a separate PR

Thank you for your suggestion. I have created a separate PR for the second commit that fixes the OSGiShellCommand test error. Here is the link to the new PR: [Fix OSGiShellCommand Test Error](https://github.com/eclipse-ee4j/glassfish/pull/24449)

I appreciate your guidance and look forward to further feedback.

avpinchuk commented on nucleus/common/common-util/src/test/java/com/sun/enterprise/universal/process/ProcessManagerTest.java:
```diff
@@ -73,4 +76,76 @@ private List<String> hugeInput() {
+            assertEquals(0, exitCode);
+            assertEquals("hello\n", pm.getStdout());
+        } catch (ProcessManagerException e) {
+            fail("Process execution failed");
```
I think, it's better to use new JUnit Jupiter API, i.e. `assertAll`, `assertDoesNotThrow` and `assertThrows` instead of `fail`. E.g.,

```java
...
int exitCode = assertDoesNotThrow(pm::execute);
assertAll(
    () -> assertEquals(0, exitCode),
    () -> assertEquals("hello\n", pm.getStdout()
);
...
```

In my opinion this make a code more clean.

avpinchuk commented on nucleus/common/common-util/src/test/java/com/sun/enterprise/universal/process/ProcessManagerTest.java:
```diff
@@ -73,4 +76,76 @@ private List<String> hugeInput() {
+            assertEquals(0, exitCode);
+            assertEquals(value + "\n", pm.getStdout());
+        } catch (ProcessManagerException e) {
+            fail("Process execution failed");
```
Here too

avpinchuk commented on nucleus/common/common-util/src/test/java/com/sun/enterprise/universal/process/ProcessManagerTest.java:
```diff
@@ -73,4 +76,76 @@ private List<String> hugeInput() {
+            String expectedOutput = String.join("\n", inputLines) + "\n";
+            assertEquals(expectedOutput, pm.getStdout());
+        } catch (ProcessManagerException e) {
+            fail("Process execution failed");
```
And here

avpinchuk commented on nucleus/common/common-util/src/test/java/com/sun/enterprise/universal/process/ProcessManagerTest.java:
```diff
@@ -73,4 +76,76 @@ private List<String> hugeInput() {
+    }
+
+    @Test
+    void testSetTimeoutMsec_LessThanExecutionTime() {
```
Maybe it be better to rename to `testSetTimeoutLessThanExecutionTime`?

avpinchuk commented on nucleus/common/common-util/src/test/java/com/sun/enterprise/universal/process/ProcessManagerTest.java:
```diff
@@ -73,4 +76,76 @@ private List<String> hugeInput() {
+        pm.setTimeoutMsec(timeoutMsec);
+        try {
+            pm.execute();
+            fail("Expected ProcessManagerTimeoutException was not thrown");
```
Simply use

```java
assertThrows(ProcessManagerTimeoutException.class, pm::execute());
```
instead.

avpinchuk commented on nucleus/common/common-util/src/test/java/com/sun/enterprise/universal/process/ProcessManagerTest.java:
```diff
@@ -73,4 +76,76 @@ private List<String> hugeInput() {
+            int exitCode = pm.execute();
+            assertEquals(0, exitCode);  // Assert that the process completes successfully
+        } catch (ProcessManagerException e) {
+            fail("Process execution failed", e);
```
Here too

avpinchuk commented on nucleus/common/common-util/src/test/java/com/sun/enterprise/universal/process/ProcessManagerTest.java:
```diff
@@ -73,4 +76,76 @@ private List<String> hugeInput() {
+    }
+
+    @Test
+    void testSetTimeoutMsec_GreaterThanExecutionTime() {
```
Please, rename to `testSetTimeoutGreaterThanExecutionTime`.

avpinchuk commented:
Also, because you run UNIX-like commans in `pm.execute()`, maybe make sense add `@DisabledOnOs(WINDOWS)` annotation to your methods, possible, as a temporary solution?

hs536 reviewed (approved):
LGTM

avpinchuk reviewed (changes requested):
@mz1999, please look at my comments

mz1999 committed:
Refactor ProcessManager tests based on feedback

- Adopted JUnit Jupiter API for assertions
- Renamed test methods for clarity
- Added @DisabledOnOs(WINDOWS) annotation due to UNIX-like commands in tests

mz1999 commented:
> @mz1999, please look at my comments

I have made the changes according to your suggestions. 
I appreciate your detailed feedback, which has been very helpful. I've learned a lot from it.

dmatej commented:
> > @mz1999, please look at my comments
> 
> I have made the changes according to your suggestions. I appreciate your detailed feedback, which has been very helpful. I've learned a lot from it.

We appreciate your help too!

avpinchuk commented:
@mz1999, happy to help. Thank you too for your work!

pzygielo commented:
Perhaps I could get help here on following test failure.
When executed locally on linux, jdk 17 temurin, current master (f985ae7390941832b0f2f7c6a2ae651723598561):
```
[ERROR] Failures: 
[ERROR]   ProcessManagerTest.testSetStdinLines:112 Multiple Failures (1 failure)
	org.opentest4j.AssertionFailedError: expected: <line1
line2
line3
> but was: <>
[INFO] 
[ERROR] Tests run: 112, Failures: 1, Errors: 0, Skipped: 1
```

dmatej commented:
@pzygielo I have another, seems the output is not flushed.

```
Multiple Failures (1 failure)
	org.opentest4j.AssertionFailedError: expected: <line1
line2
line3
> but was: <line1
line2
>
```

pzygielo commented:
Thanks for checking.

mz1999 commented:
> 

Is it possible that the process executes too quickly, exits before the reader thread has read all the data, and closes the outputstream, resulting in incomplete data reading?

dmatej commented:
> > 
> 
> Is it possible that the process executes too quickly, exits before the reader thread has read all the data, and closes the outputstream, resulting in incomplete data reading?

Yes, that's it, but as it is a parallel thread, it simply tries to read the output and at any time it can retrieve null. Instead of the for cycle there should be rather a semaphore, `boolean stop`, `false` by defaul, with a setter. When the process would end or would be killed, the reader thread then would be asked to read the rest of the output and end.

It can also use StringBuilder instead of StringBuffer and the buffer could be moved inside the ThreadReader, then retrieved after the process finished. And the `threads` field could be replaced by direct references to both threadreaders. The thread will stop, but the instance then can still provide the output, so then you could call `threadReader.getStdOut()` after the thread stopped.

dmatej commented:
Ok, I have prepared a fix, because I wasn't sure with my answer :-)
